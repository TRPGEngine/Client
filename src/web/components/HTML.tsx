import React, { useMemo } from 'react';
import { filterXSS } from 'xss';
import { Parser } from 'html-to-react';
import { TMemo } from '@shared/components/TMemo';

const parser = new Parser();

interface Props {
  html: string;
}
const HTML: React.FC<Props> = TMemo((props) => {
  const el = useMemo(() => {
    const html = filterXSS(props.html, { css: false }); // 防止xss
    return parser.parse(html);
  }, [props.html]);

  return <div>{el}</div>;
});
HTML.displayName = 'HTML';

export default HTML;
