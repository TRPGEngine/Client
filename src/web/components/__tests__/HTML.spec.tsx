import React from 'react';
import HTML from '@web/components/HTML';
import { create } from 'react-test-renderer';

describe('HTML Component', () => {
  test.each([
    ['<h1>Test Title</h1>'],
    ['<script src="http://example.com/script.js"></script>'],
    ['<style>*{color: red;}</style>'],
  ])('render "%s"', (originText) => {
    const wrapper = create(<HTML html={originText} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
