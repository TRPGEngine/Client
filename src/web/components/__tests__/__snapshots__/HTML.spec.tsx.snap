// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Component render "<h1>Test Title</h1>" 1`] = `
<div>
  <h1>
    Test Title
  </h1>
</div>
`;

exports[`HTML Component render "<script src="http://example.com/script.js"></script>" 1`] = `
<div>
  &lt;script src="http://example.com/script.js"&gt;&lt;/script&gt;
</div>
`;

exports[`HTML Component render "<style>*{color: red;}</style>" 1`] = `
<div>
  &lt;style&gt;*{color: red;}&lt;/style&gt;
</div>
`;
